pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,453 @@
+      <https://quantum-journal.org/papers/q-2020-05-11-263/>`__ (2020).
+
+    **Example**
+
+    For VQE/VQE-like problems, the objective function for the optimizer can be
+    realized as an :class:`~.ExpvalCost` object, constructed using a :class:`~.Hamiltonian`.
+
+    >>> coeffs = [2, 4, -1, 5, 2]
+    >>> obs = [
+    ...   qml.PauliX(1),
+    ...   qml.PauliZ(1),
+    ...   qml.PauliX(0) @ qml.PauliX(1),
+    ...   qml.PauliY(0) @ qml.PauliY(1),
+    ...   qml.PauliZ(0) @ qml.PauliZ(1)
+    ... ]
+    >>> H = qml.Hamiltonian(coeffs, obs)
+    >>> dev = qml.device("default.qubit", wires=2, analytic=False)
+    >>> cost = qml.ExpvalCost(qml.templates.StronglyEntanglingLayers, H, dev)
+
+    Once constructed, the cost function can be passed directly to the
+    optimizer's ``step`` function:
+
+    >>> params = qml.init.strong_ent_layers_uniform(n_layers=2, n_wires=2)
+    >>> opt = qml.ShotAdaptiveOptimizer(min_shots=10)
+    >>> for i in range(60):
+    ...    params = opt.step(cost, params)
+    ...    print(f"Step {i}: cost = {cost(params)}, shots_used = {opt.shots_used}")
+    Step 0: cost = -5.686, shots_used = 240
+    Step 1: cost = -2.983999999999999, shots_used = 336
+    Step 2: cost = -4.974, shots_used = 624
+    Step 3: cost = -5.534, shots_used = 1054
+    Step 4: cost = -6.5, shots_used = 1798
+    Step 5: cost = -6.684, shots_used = 2942
+    Step 6: cost = -6.992, shots_used = 4350
+    Step 7: cost = -6.970000000000001, shots_used = 5814
+    Step 8: cost = -6.998, shots_used = 7230
+    Step 9: cost = -6.686, shots_used = 9006
+    Step 10: cost = -6.85, shots_used = 11286
+    Step 11: cost = -6.628000000000001, shots_used = 14934
+    Step 12: cost = -6.862, shots_used = 17934
+    Step 13: cost = -7.188, shots_used = 22950
+    Step 14: cost = -6.9879999999999995, shots_used = 28302
+    Step 15: cost = -7.380000000000001, shots_used = 34134
+    Step 16: cost = -7.656, shots_used = 41022
+    Step 17: cost = -7.212, shots_used = 48918
+    Step 18: cost = -7.532, shots_used = 56286
+    Step 19: cost = -7.462000000000001, shots_used = 63822
+    Step 20: cost = -7.306000000000001, shots_used = 72534
+    Step 21: cost = -7.234, shots_used = 82014
+    Step 22: cost = -7.314, shots_used = 92838
+    """
+
+    def __init__(self, min_shots, weighted_random_sampling=True, mu=0.99, b=1e-6, stepsize=0.07):
+        self.wrs = weighted_random_sampling
+        self.trainable_args = set()
+
+        # hyperparameters
+        self.min_shots = min_shots
+        self.mu = mu  # running average constant
+        self.b = b  # regularization bias
+        self.lipschitz = None
+
+        # keep track of the total number of shots used
+        self.shots_used = 0
+        # total number of iterations
+        self.k = 0
+        # Number of shots per parameter
+        self.s = None
+        # maximum number of shots allowed per iteration
+        self.max_shots = None

```

Comment:
Here it sounds like `self.max_shots` should be a user-provided parameter to cap out device executions, but doesn't seem to be how it is used.  Reading the code, it seems to be the maximum number actually used on one parameter in that iteration.

Reply:
Oops, good catch. Have updated the comment to reflect the actual behaviour of this attribute!